### PR TITLE
FOGL-8040: Initial design of API Control Flow & API integration testing

### PR DESCRIPTION
**Pull request by monika-Dianomic**

### CODE REVIEW

File: src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+    selectedDestinationType = { cpdid: 0, name: "Broadcast" };  // Typecast with DestinationType enum
+    selectedDestinationName = null;
+ 
+    destinationTypes = [];
+    destinationNames = [];
+
+    editMode = false;
+    allowExecute = false;
+    
+    _name: string;
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts:48]: may be say it entrypointName or apiFlowName

File: src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+    af: APIFlow;
+    apiFlowForm: FormGroup;
+    apiFlowFormCopy: FormGroup;
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts:51]: Why do we need a copy ?

File: src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+
+    allUsers: User[];
+    selectedUsers: [];
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts:54]: this is hold local data on selection?

File: src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+            this.ngProgress.done();
+            if (error.status === 0) {
+              console.log('service down ', error);
+            } else {
+              this.alertService.error(error.statusText);
+            }
+          });
+    }
+
+    fillParameters(param, controlType) {
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts:184]: keep this near add parameter, init parameters

File: src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts [hunk 5]
```diff
@@ -0,0 +1,526 @@
...
+            }
+            this.getSelectedType();
+            
+            // TODO: FOGL-8070
+            this.af.anonymous = data.anonymous === 't' || data.anonymous === true ? true : false;
+            
+            let v = <FormArray>this.apiFlowForm.controls['variables'];
+            v.clear();
+            let c = <FormArray>this.apiFlowForm.controls['constants'];
+            c.clear();
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts:169]: May be move this to fillParameter and wrap in if variable if constants

File: src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+
+    fillParameters(param, controlType) {
+        let i = 0
+        for (const [key, value] of Object.entries(param)) {
+            this.addParameter({ index: i, key: key, value: value }, controlType);
+            i++;
+        }
+    }
+
+    addAPIFlow(data) {
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts:192]: Please keep statements in order of form control

Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts:192]: name this as saveAPIFlow

File: src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts [hunk 7]
```diff
@@ -0,0 +1,526 @@
...
+        payload.type = this.selectedType;
+        payload.destination = destination;     
+        if (destination !== 'broadcast') {
+            payload[destination] = this.selectedDestinationName;
+        }
+        if (this.editMode) {
+            this.updateAPIFlow();
+            return;
+        }
+        this.controlAPIFlowService.createAPIFlow(payload)
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts:222]: move to this.addAPIFlow

File: src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts [hunk 8]
```diff
@@ -0,0 +1,526 @@
...
+        } else {
+            delete payload.constants;
+        }
+
+        payload.type = this.selectedType;
+        payload.destination = destination;     
+        if (destination !== 'broadcast') {
+            payload[destination] = this.selectedDestinationName;
+        }
+        if (this.editMode) {
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts:218]: Add a newline above

File: src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts [hunk 9]
```diff
@@ -0,0 +1,526 @@
...
+            this.ngProgress.done();
+            if (error.status === 0) {
+              console.log('service down ', error);
+            } else {
+              this.alertService.error(error.statusText);
+            }
+          });
+    }
+
+    public toggleDropDown(id: string) {
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts:261]: move it at other place, keep functions def logically grouped

File: src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts [hunk 10]
```diff
@@ -0,0 +1,526 @@
...
+            this.ngProgress.done();
+            if (error.status === 0) {
+                console.log('service down ', error);
+            } else {
+                this.alertService.error(error.statusText);
+            }
+        });
+    }
+
+    public toggleDropdown(id) {
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts:314]: is this duplicate function?

File: src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts [hunk 11]
```diff
@@ -0,0 +1,526 @@
...
+            this.ngProgress.done();
+            if (error.status === 0) {
+              console.log('service down ', error);
+            } else {
+              this.alertService.error(error.statusText);
+            }
+          });
+    }
+
+    removeParameter(index: number, param) {
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts:481]: move near addParameter

File: src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts [hunk 12]
```diff
@@ -0,0 +1,526 @@
...
+
+    getFormControls(type): AbstractControl[] {
+        return (<FormArray>this.apiFlowForm.get(type)).controls;
+    }
+
+    selectDestinationName(value) {
+        this.selectedDestinationName = value === 'Select Destination Name' ? null : value;
+    }
+
+    getDestinationNames(selectedType) {
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts:364]: this function cannot be used from a component ts? Many functions here are already in use in Control pipeline and scripts, users

File: src/app/components/core/control-dispatcher/api-flow/api-flow.component.ts [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+        });
+    }
+
+    // Set Anonymous T/F
+    onCheckboxClicked(event, name) {
+    console.log(event.target.checked, name)
+    // call update
+    } 
+
+    // openModal(name, key, message) {
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/api-flow.component.ts:169]: cleanup?

File: src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts [hunk 14]
```diff
@@ -0,0 +1,526 @@
...
+            operation_name: '',
+            destination: 'broadcast',
+            constants: {},
+            variables: {},
+            anonymous: false,
+            allow: []
+          };
+        this.getDestTypes();
+        this.getUsers();
+        this.types = Object.keys(APIFlowType).map(key => APIFlowType[key]).filter(value => typeof value === 'string') as string[];
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts:104]: break it into multilines?

File: src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts [hunk 15]
```diff
@@ -0,0 +1,526 @@
...
+              this.ngProgress.done();
+              if (error.status === 0) {
+                console.log('service down ', error);
+              } else {
+                this.alertService.error(error.statusText);
+              }
+            });
+    }
+
+    selectAllowedUsers(user) {
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts:515]: arg should be usernames. (plural)

File: src/app/components/core/control-dispatcher/api-flow/api-flow.component.html [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+<app-confirmation-dialog id="confirmation-anonymous-dialog">                           
+    <header class="modal-card-head">
+        <p class="modal-card-title is-size-6">{{allowAnonymous ? 'Allow Anonymous Access' : 'Allow Specific Authenticated Users'}}</p>
+        <button type="button" class="delete" aria-label="close" (click)="closeModal('confirmation-anonymous-dialog')"></button>
+    </header>
+    <section class="modal-card-body anonymous-modal">
+        <ng-container *ngIf="allowAnonymous">
+            <article *ngIf="allowAnonymous" class="message my-2">
+                <div class="message-body pl-3 py-2"> 
+                    <small class="anonymous-warning">This will allow any user to execute the entrypoint. Its recommended to allow authenticated users only.</small>
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/api-flow.component.html:178]: class anonymous-warning need to be added to css file?

File: src/app/components/core/control-dispatcher/api-flow/api-flow.component.html [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+                            </ng-container>
+                        </ng-container>
+                    </div>
+                </div>
+            </div>
+        </form>
+    </section>
+    <footer class="modal-card-foot">
+        <button class="button is-small" type="button" (click)="closeModal('confirmation-execute-dialog')">Cancel</button>
+        <button class="button is-small" type="button" (click)="requestAPIFlow(apiFlowForm.value)">Execute</button>
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/api-flow.component.html:151]: Fix button color please as it was.

File: src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.html [hunk 18]
```diff
@@ -0,0 +1,483 @@
...
+</footer>
+</app-confirmation-dialog>
+
+<app-confirmation-dialog id="confirmation-delete-dialog">
+    <header class="modal-card-head">
+        <p class="modal-card-title is-size-6">Delete</p>
+        <button type="button" class="delete" aria-label="close" (click)="closeModal('confirmation-delete-dialog')"></button>
+    </header>
+    <section class="modal-card-body">
+        Are you sure, you want to delete the Entry Point <b>{{this.af.name}}</b>?
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.html:477]: use `You` after ,

File: src/app/components/core/control-dispatcher/api-flow/api-flow.component.html [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+<app-confirmation-dialog id="confirmation-anonymous-dialog">                           
+    <header class="modal-card-head">
+        <p class="modal-card-title is-size-6">{{allowAnonymous ? 'Allow Anonymous Access' : 'Allow Specific Authenticated Users'}}</p>
+        <button type="button" class="delete" aria-label="close" (click)="closeModal('confirmation-anonymous-dialog')"></button>
+    </header>
+    <section class="modal-card-body anonymous-modal">
+        <ng-container *ngIf="allowAnonymous">
+            <article *ngIf="allowAnonymous" class="message my-2 is-warning">
+                <div class="message-body pl-3 py-2"> 
+                    <small class="anonymous-warning">This will allow any user to execute the entrypoint. Its recommended to allow authenticated users only.</small>
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/api-flow.component.html:178]: This action .....entry point. It's .... only.

File: src/app/components/core/control-dispatcher/api-flow/api-flow.component.ts [hunk 20]
```diff
@@ -0,0 +1,246 @@
...
+    templateUrl: './api-flow.component.html',
+    styleUrls: ['./api-flow.component.css']
+})
+
+export class APIFlowComponent implements OnInit {
+    apiFlows = []; // TODO: typecast APIFlow
+    epName: string = '';
+    allUsers: User[];
+    allowedUser: string[];
+    allowAnonymous: boolean;
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/api-flow.component.ts:26]: this is for allowExecute as we have in details page?

File: src/app/components/core/control-dispatcher/api-flow/api-flow.component.ts [hunk 21]
```diff
@@ -0,0 +1,246 @@
...
+@Component({
+    selector: 'app-api-flow',
+    templateUrl: './api-flow.component.html',
+    styleUrls: ['./api-flow.component.css']
+})
+
+export class APIFlowComponent implements OnInit {
+    apiFlows = []; // TODO: typecast APIFlow
+    epName: string = '';
+    allUsers: User[];
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/api-flow.component.ts:24]: Add a line break before this, to logically group the member variables & better readability.

File: src/app/components/core/control-dispatcher/api-flow/api-flow.component.ts [hunk 22]
```diff
@@ -0,0 +1,246 @@
...
+    allowedUser: string[];
+    allowAnonymous: boolean;
+
+    loggedInUsername: string;
+    apiFlowForm: FormGroup;
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/api-flow.component.ts:29]: what is the use of this form in list page?

File: src/app/components/core/control-dispatcher/api-flow/api-flow.component.ts [hunk 23]
```diff
@@ -0,0 +1,246 @@
...
+            this.getAPIFlows();
+            // TODO: change checkbox status on cancel
+            // this.allowAnonymous = !this.allowAnonymous;
+            // this.allowedUser = [];
+        }
+        this.dialogService.close(id);
+    }
+
+    goToLink(urlSlug: string) {
+    this.docService.goToSetPointControlDocLink(urlSlug);
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/api-flow.component.ts:243]: indent

File: src/app/models/api-flow.ts [hunk 24]
```diff
@@ -0,0 +1,19 @@
...
+  constants: {};
+  variables: {};
+  anonymous: any;  // TODO: FOGL-8070
+  allow: string[] // List of allowed users' username
+}
+
+export enum APIFlowType {
+  write = 0,
+  operation = 1
+}
```
[the remark this answers is not in the document]
Comment (praveen-garg) [reply src/app/models/api-flow.ts:19]: I'm okay if that simplifies the stuff in terms of dropdown binding and selection etc. However I see no harm to as there can only be one of these value at a time & to have a enum of type string.

File: src/app/components/core/control-dispatcher/api-flow/api-flow.component.html [hunk 25]
```diff
@@ -0,0 +1,211 @@
...
+                    <th></th>
+                </tr>
+                </thead>
+                <tbody>
+                    <tr *ngFor="let ep of apiFlows">
+                        <td class="align-content"> 
+                            <a class="button is-text ep-name mx-0 px-0" [routerLink]="['/control-dispatcher/api', ep?.name]" title={{ep.description}}>{{ep.name}}</a>
+                        </td>
+                        <td class="align-content">
+                            <span class="is-small"><!--{{ep.type | titlecase}}-->Operation</span>
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/api-flow.component.html:38]: Add a TODO

File: src/app/components/core/control-dispatcher/api-flow/api-flow.component.html [hunk 26]
```diff
@@ -0,0 +1,211 @@
...
+                            <span class="help sub">
+                                <!--{{ep.type == 'operation' ? ep.operation_name : ''}} -->
+                                Off
+                            </span>
+                        </td>
+                        <td></td>
+                        <td class="align-content">
+                            <span class="is-small"><!--{{ep.destination}} --> Any</span>
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/api-flow.component.html:46]: Add a TODO

File: src/app/components/core/control-dispatcher/api-flow/api-flow.component.html [hunk 27]
```diff
@@ -0,0 +1,211 @@
...
+                            <div class="dropdown-menu" id="dropdown-menu" role="menu">
+                            <div class="dropdown-content">
+                                <!-- TODO: FOGL-8037
+                                Add check of logged-in user vs allowed users (Editor or above role) if not anonymous 
+                                Show modal to override variable values
+
+                                There is no enough info on list page. So, only checked for permitted
+                                -->
+                            
+                            <a class="dropdown-item" *ngIf="ep.permitted" (click)="getAPIFlow(ep.name);openModal('confirmation-execute-dialog', ep.name, ep.description)">
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/api-flow.component.html:83]: This check is WRONG!

We must check for each record the same way, we have in details page for this.allowExecute, i.e. either anonymous or loggedInUser? exists in allow list. Add todo what if an Admin or, below the role of Editor.

File: src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts [hunk 28]
```diff
@@ -0,0 +1,512 @@
...
+            if (error.status === 0) {
+              console.log('service down ', error);
+            } else {
+              this.alertService.error(error.statusText);
+            }
+          });
+    }
+
+    checkAndRequestAPIFlow() {
+      // TODO: FOGL-8079 (blank values for variables are not allowed)
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts:286]: For now add a check after this 
```
 if name = "check_with_no_vars"
    then unset af.vars locally
    and let flow continue
```

File: src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts [hunk 29]
```diff
@@ -0,0 +1,512 @@
...
+    selectDestinationName(value) {
+        this.selectedDestinationName = value === 'Select Destination Name' ? null : value;
+    }
+
+    getDestinationNames(selectedType) {
+        this.destinationNames = [];
+        this.selectedDestinationName = null;
+        this.af.destination = selectedType.name === 'Select Destination Type' ? '' : selectedType.name;
+        switch (selectedType.name) {
+          case 'Any':
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/add-edit-api-flow/add-edit-api-flow.component.ts:362]: remove Any; 

Though suggest to use existing  function in control pipeline or move such to common /cc @monika-Dianomic

**Review by praveen-garg — APPROVED** [review]

**Comment by monika-Dianomic** [comment]
This PR is for initial design only; refactoring will be done once Backend is ready and issues are fixed.

**Review by praveen-garg — APPROVED** [review]

File: src/styles.css [hunk 30]
```diff
@@ -382,4 +382,15 @@ a.dropdown-item:hover {
...
   }
+
+  .description-text {
+    overflow: hidden;
+    text-overflow: ellipsis;
+    white-space: nowrap;
+    max-width: 32rem
+  }
+
+  #confirmation-execute-dialog .modal-card-body {
```
Comment (praveen-garg) [inline src/styles.css:393]: Avoid these in styles.css?

File: src/app/models/index.ts [hunk 31]
```diff
@@ -1 +1,2 @@
 export * from './user';
+export * from './api-flow';
```
Comment (praveen-garg) [inline src/app/models/index.ts:2]: should be under control/

File: src/app/components/core/control-dispatcher/api-flow/api-flow.component.ts [hunk 32]
```diff
@@ -0,0 +1,248 @@
...
+    setEdit(name, state){
+      this.editMode["name"] = name;
+      this.editMode["edit"] = state;
+      if (state == false){
+        this.updateDescription();
+      }
+    }
+
+    resetEditMode() {
+      this.editMode = {name: null, edit: false, newVal: null}
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/api-flow.component.ts:197]: newVal should be _value_, we are setting _value_ in  descriptionChange and using that in updateDescription

File: src/app/components/core/control-dispatcher/api-flow/api-flow.component.css [hunk 33]
```diff
@@ -0,0 +1,39 @@
...
+    border: none;
+}
+
+.anonymous-modal {
+    overflow: visible;
+}
+
+.anonymous-modal-text {
+    font-size: 0.9rem;
+}
```
Comment (praveen-garg) [inline src/app/components/core/control-dispatcher/api-flow/api-flow.component.css:11]: required?

**Pull request merged**